sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala

```diff
@@ -129,6 +163,45 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
               child = Project(innerProjectList.toSeq, child)
             )
           }
+
+        case agg @ Aggregate(groupingExpressions, aggregateExpressions, _) if agg.resolved
+            && aggregateExpressions.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+
+          val newAggExprs = collection.mutable.Set.empty[NamedExpression]
+          val expressionMap = collection.mutable.LinkedHashMap.empty[Expression, NamedExpression]
+          val projectExprs = aggregateExpressions.map { exp =>
+            exp.transformDown {
+              case aggExpr: AggregateExpression =>
+                // Doesn't support referencing a lateral alias in aggregate function
+                if (aggExpr.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) {
+                  aggExpr.collectFirst {
+                    case lcaRef: LateralColumnAliasReference =>
+                      throw QueryCompilationErrors.lateralColumnAliasInAggFuncUnsupportedError(
+                        lcaRef.nameParts, aggExpr)
+                  }
+                }
+                val ne = expressionMap.getOrElseUpdate(aggExpr.canonicalized, assignAlias(aggExpr))
+                newAggExprs += ne
```

Comment:
So we are extracting all AggregateExpression and grouping expressions from `aggregateExpressions`. Shall we just extract the ones which are required by the lateral column aliases?

Reply:
No, actually all the grouping expression and aggregate expressions in the aggregate list should be extracted,  otherwise they will be unresolved from the upper Project, correct?

Reply:
So for the following query
```
create table table_1(a int, b int, c int);
select a as x, x+1, sum(b)+sum(c) from table_1 group by a;
```
Do we need to extract `sum(b)` and `sum(c)` and put them separately in the new aggregation expressions?

Reply:
Yes we do. This is the general strategy of this design, extract **all** aggregation functions or grouping expressions. It should be simple and avoid many extra considerations of code path (e.g., mark if the alias is used as lateral alias by later expressions).